qiskit_experiments/library/randomized_benchmarking/mirror_rb_experiment.py

```diff
@@ -0,0 +1,522 @@
+
+        Additional information:
+            The circuits are constructed iteratively; each circuit is obtained
+            by extending the previous circuit (without the inversion and measurement gates)
+        """
+        qubits = list(range(self.num_qubits))
+        circuits = []
+
+        circs = [QuantumCircuit(self.num_qubits) for _ in range(len(lengths))]
+
+        for current_length, group_elt_circ in enumerate(elements[: (len(elements) // 2)]):
+            if isinstance(group_elt_circ, tuple):
+                group_elt_gate = group_elt_circ[0]
+            else:
+                group_elt_gate = group_elt_circ
+
+            if not isinstance(group_elt_gate, Instruction):
+                group_elt_gate = group_elt_gate.to_instruction()
+            for circ in circs:
+                circ.barrier(qubits)
+                circ.append(group_elt_gate, qubits)
+
+            double_current_length = (
+                (current_length + 1) * 2 + 1 if len(elements) % 2 == 1 else (current_length + 1) * 2
+            )
+            if double_current_length in lengths:
+                rb_circ = circs.pop()
+                inv_start = (
+                    (-(current_length + 1) - 1) if len(elements) % 2 == 1 else -(current_length + 1)
+                )
+                for inv in elements[inv_start:]:
+                    if isinstance(inv, tuple):
+                        group_elt_gate = inv[0]
+                    else:
+                        group_elt_gate = inv
+
+                    if not isinstance(group_elt_gate, Instruction):
+                        group_elt_gate = group_elt_gate.to_instruction()
+                    rb_circ.barrier(qubits)
+                    rb_circ.append(group_elt_gate, qubits)
+                rb_circ.metadata = {
+                    "experiment_type": self._type,
+                    "xval": double_current_length,
+                    "group": "Clifford",
+                    "physical_qubits": self.physical_qubits,
+                    "target": self._clifford_utils.compute_target_bitstring(rb_circ),
+                    "inverting_pauli_layer": self._inverting_pauli_layer,
+                }
+                rb_circ.measure_all()
+                circuits.append(rb_circ)
+        return circuits
+
+
+class MirrorRBPyGSTi(MirrorRB):
+    """Mirror RB experiment that uses pyGSTi's circuit generation. This subclass
+    is primarily used for testing."""
+
+    def __init__(
+        self,
+        qubits: Sequence[int],
+        lengths: Iterable[int],
+        local_clifford: bool = True,
+        pauli_randomize: bool = True,
+        two_qubit_gate_density: float = 0.2,
+        backend: Optional[Backend] = None,
+        num_samples: int = 3,
+        seed: Optional[Union[int, SeedSequence, BitGenerator, Generator]] = None,
+        full_sampling: bool = False,
+        inverting_pauli_layer: bool = False,
+    ):
+        """Initialize a mirror randomized benchmarking experiment that uses
```

Comment:
What is the benefit of this class for users? If this class is just for validation, you should extract the circuit generation logic and use it only in the unittest to compare the output circuits. If PyGSTi gives you something like x10 performance improvement, perhaps it's worth having it here, but not as a subclass. You can refer to the implementation of tomography module. There are several fitters available (numpy, scipy, cvxopt) but these are the single class and the internal logic is dynamically switched based on the analysis option. Something like this would work.

Reply:
We plan to make a separate PR for moving this class to the internal repository.